### PR TITLE
Proof Presentation Request Initial Commit

### PR DESCRIPTION
**Pull request by ankita-p17**
This includes functionality to create a proof presentation request and send it to the intended user.
This is a work in progress PR.

### CODE REVIEW

File: src/lib/modules/PresentationModule.ts [hunk 1]
```diff
@@ -0,0 +1,24 @@
+import { ConnectionRecord } from '../storage/ConnectionRecord';
+import { createOutboundMessage } from '../protocols/helpers';
+import { MessageSender } from '../agent/MessageSender';
+import { presenationRequestTemplate, PresentationService } from '../protocols/presentproof/PresentationService';
+
+export class CredentialsModule {
```
Comment (jakubkoci) [inline src/lib/modules/PresentationModule.ts:6]: Don't forget to rename `CredentialsModule`. I'm thinking if the name `Proof` wouldn't be better. So `ProofsModule`, `ProofService`, `ProofRecord`, ... I looked at .NET and they also use this name.

File: src/lib/protocols/presentproof/PresentationStates.ts [hunk 2]
```diff
@@ -0,0 +1,10 @@
+export enum PresentationStates {
+    PROPOSAL_SENT = "proposal_sent",
+    PROPOSAL_RECEIVED = "proposal_received",
+    REQUEST_SENT = "request_sent",
+    REQUEST_RECEIVED = "request_received",
+    PRESENTATION_SENT = "presentation_sent",
+    PRESENTATION_RECEIVED = "presentation_received",
+    VERIFIED = "verified",
+    PRESENTATION_ACKED = "presentation_acked"
+}
```
Comment (jakubkoci) [inline src/lib/protocols/presentproof/PresentationStates.ts:10]: We're using [TypeScript String enums](https://www.typescriptlang.org/docs/handbook/enums.html#string-enums) with the following convention:

```
export enum ConnectionState {
  Init = 'INIT',
  Invited = 'INVITED',
  Requested = 'REQUESTED',
  Responded = 'RESPONDED',
  Complete = 'COMPLETE',
}
```

File: src/lib/protocols/presentproof/messages/ProofRequest.ts [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+
+    public constructor(proofRequest: ProofRequestInterface) {
+        if (proofRequest) {
+            this.name = proofRequest.name;
+            this.version = proofRequest.version;
+            this.nonce = proofRequest.nonce;
+            this.requestedAttributes = proofRequest.requestedAttributes;
+            this.requestedPredicates = proofRequest.requestedPredicates;
+        }
+    }
```
Comment (jakubkoci) [inline src/lib/protocols/presentproof/messages/ProofRequest.ts:13]: Please check the formatting of all files with `yarn check-format` eventually fix formatting with `yarn format`.

**Review by jakubkoci — CHANGES_REQUESTED** [review]
Thanks, @ankita-p17 for the contribution. It looks like a good starting point. It would be nice to have some e2e test for a small portion of the functionality if possible. I've prepared a short video explanation of [Test in Aries JS framework](https://www.youtube.com/watch?v=__CbgqTPTZs), but we can have a call and I'm happy to help you with that if you need.

File: src/lib/modules/PresentationModule.ts (hunk 1 above)
Comment (ankita-p17) [reply src/lib/modules/PresentationModule.ts:6]: Sure, will update that. Thanks for your input. I will update the test as well accordingly

File: src/lib/protocols/presentproof/PresentationStates.ts (hunk 2 above)
Comment (ankita-p17) [reply src/lib/protocols/presentproof/PresentationStates.ts:10]: Sure, will update. Thanks

File: src/lib/protocols/presentproof/messages/Attachment.ts [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+
+interface AttachmentOptions {
+  id?: string;
+  mimeType: string;
+  data: {
+    base64: string;
+  };
+}
+
+export class Attachment {
```
Comment (TimoGlastra) [inline src/lib/protocols/presentproof/messages/Attachment.ts:13]: Is the Attachment for the present-proof protocol different from the issue-credential protocol? Maybe we could reuse the attachment class and make it a general purpose Attachment decorator per [Aries RFC 0017: Attachments](https://github.com/hyperledger/aries-rfcs/blob/master/concepts/0017-attachments/README.md)

File: src/lib/protocols/presentproof/messages/MessageType.ts [hunk 5]
```diff
@@ -0,0 +1,7 @@
+export enum MessageType {
+    PresentationRequest = 'did:sov:BzCbsNYhMrjHiqZDTUASHg;spec/present-proof/1.0/propose-presentation',
```
Comment (TimoGlastra) [inline src/lib/protocols/presentproof/messages/MessageType.ts:2]:
```suggestion
    PresentationRequest = 'did:sov:BzCbsNYhMrjHiqZDTUASHg;spec/present-proof/1.0/request-presentation',
```

File: src/lib/protocols/presentproof/messages/Attachment.ts [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+    }
+  }
+
+  @Expose({ name: '@id' })
+  public id!: string;
+
+  @Expose({ name: 'mime-type' })
+  public mimeType!: string;
+
+  @Expose({ name: 'data' })
```
Comment (TimoGlastra) [inline src/lib/protocols/presentproof/messages/Attachment.ts:28]: Expose is not needed if json key is the same as the class property

File: src/lib/protocols/presentproof/messages/ProofRequest.ts [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+
+    @Expose({ name: 'name' })
+    public name!: string;
+    @Expose({ name: 'version' })
+    public version!: string;
+    @Expose({ name: 'nonce' })
+    public nonce!: string;
```
Comment (TimoGlastra) [inline src/lib/protocols/presentproof/messages/ProofRequest.ts:20]: Same here. `@Expose` is only needed if the json property name is different from the class property name

**Review by TimoGlastra — COMMENTED** [review]
Thanks for the PR @ankita-p17 :)

**Comment by TimoGlastra** [comment]
This is implemented as part of #154

**Pull request closed without merging**